strings/base_macros.h

```diff
@@ -77,14 +77,53 @@ struct IUnknown;
-// this type does not exist in C++17 we need to use a macro to optionally add parameters and forwarding wherever appropriate.
+#if defined(__cpp_consteval)
+#define WINRT_IMPL_CONSTEVAL consteval
+#else
+#define WINRT_IMPL_CONSTEVAL constexpr
+#endif
+
+// The intrinsics (such as __builtin_FILE()) that power std::source_location are also used to power winrt:impl::slim_source_location.
+// The source location needs to be for the calling code, not cppwinrt itself, so that it is useful to developers building on top of
+// this library.  As a result any public-facing method that can result in an error needs a default-constructed slim_source_location
+// argument so that it will collect source information from the application code that is calling into cppwinrt.
 //
-// Some projects may decide to disable std::source_location support to prevent source code information from ending up in their
-// release binaries, or to reduce binary size.  Defining WINRT_NO_SOURCE_LOCATION will prevent this feature from activating.
-#if defined(__cpp_lib_source_location) && !defined(WINRT_NO_SOURCE_LOCATION)
+// We do not directly use std::source_location for two reasons:
+//  1) std::source_location::function_name() is unavoidable.  These strings end up in the final binary, bloating their size.  This
+//     is particularly impactful for code bases that use templates heavily.  Cases of 50% binary size growth have been observed.
+//  2) std::source_location is a cpp20 feature, which is above the cpp17 feature floor for cppwinrt.  By defining our own version
+//     we can avoid ODR violations in mixed cpp17/cpp20 builds.  cpp17 callers will have an ABI that matches cpp20 callers (they
+//     will just not have useful file/line/function information).
+//
+// Some projects may decide that the source information binary size impact is not worth the benefit.  Defining WINRT_NO_SOURCE_LOCATION
+// will prevent this feature from activating.  The slim_source_location type will be forwarded around but it will not include any
+// nonzero data.  That eliminates the biggest source of binary size overhead.
+//
+// To help with debugging the __builtin_FUNCTION() intrinsic will be used in _DEBUG builds.  This will provide a bit more diagnostic
+// value at the cost of binary size.  The assumption is that binary size is considered less important in debug builds so this tradeoff
+// is acceptable.
+
+#if !defined(__cpp_lib_source_location)
+// cpp17 mode.  The source_location intrinsics are not available.
+#define WINRT_IMPL_BUILTIN_LINE 0
+#define WINRT_IMPL_BUILTIN_FILE nullptr
+#define WINRT_IMPL_BUILTIN_FUNCTION nullptr
+#elif defined(WINRT_NO_SOURCE_LOCATION)
+// The caller has disabled source_location support.  Ensure that there is no binary size overhead for line/file/function.
+#define WINRT_IMPL_BUILTIN_LINE 0
+#define WINRT_IMPL_BUILTIN_FILE nullptr
+#define WINRT_IMPL_BUILTIN_FUNCTION nullptr
```

Comment:
Combining these identical cases is clearer, IMO:
```suggestion
#if !defined(__cpp_lib_source_location) || defined(WINRT_NO_SOURCE_LOCATION)
// Either the source_location intrinsics are not available, or the user doesn't want them.
// Ensure that there is no binary size overhead for line/file/function.
#define WINRT_IMPL_BUILTIN_LINE 0
#define WINRT_IMPL_BUILTIN_FILE nullptr
#define WINRT_IMPL_BUILTIN_FUNCTION nullptr
```